File: Components/ASP.NET/VB/ASPxSpreadsheetDemos/App_Code/DemoHelpers/GlobalCustomFunctions/DiscountFunction.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Web
Imports DevExpress.Spreadsheet.Functions

Public Class DiscountFunction
	Implements ICustomFunction
	Private Const functionName As String = "Discount"
	Private ReadOnly functionParameters() As ParameterInfo

	Public Sub New()
		Me.functionParameters = New ParameterInfo() { New ParameterInfo(ParameterType.Value, ParameterAttributes.Required), New ParameterInfo(ParameterType.Value, ParameterAttributes.Required)}
	End Sub

	Private ReadOnly Property Name() As String Implements IFunction.Name
		Get
			Return functionName
		End Get
	End Property
	Private ReadOnly Property Parameters() As ParameterInfo() Implements IFunction.Parameters
		Get
			Return functionParameters
		End Get
	End Property
	Private ReadOnly Property ReturnType() As ParameterType Implements IFunction.ReturnType
		Get
			Return ParameterType.Value
		End Get
	End Property
	' Reevaluate cells on every recalculation. 
	' The SpreadsheetControl ignores this value and assumes that the Volatile is always true.
	Private ReadOnly Property Volatile() As Boolean Implements IFunction.Volatile
		Get
			Return True
		End Get
	End Property

	Private Function Evaluate(ByVal parameters As IList(Of ParameterValue), ByVal context As EvaluationContext) As ParameterValue Implements IFunction.Evaluate
		Dim quantity As Double
		Dim price As Double
		Dim quantityParameter As ParameterValue
		Dim priceParameter As ParameterValue

		quantityParameter = parameters(0)
		If quantityParameter.IsError Then
			Return quantityParameter
		Else
			quantity = quantityParameter.NumericValue
		End If

		priceParameter = parameters(1)
		If priceParameter.IsError Then
			Return priceParameter
		Else
			price = priceParameter.NumericValue
		End If

		If quantity >= 50 Then
			Return Math.Round(quantity * price * 0.25, 2)
		End If
		If quantity >= 20 Then
			Return Math.Round(quantity * price * 0.1, 2)
		End If
		Return 0

	End Function
	Private Function GetName(ByVal culture As System.Globalization.CultureInfo) As String Implements IFunction.GetName
		Return functionName
	End Function
End Class

File: Components/ASP.NET/CS/MVCxSpreadsheetDemos/Models/GlobalCustomFunctions/DiscountFunction.cs
using System;
using System.Collections.Generic;
using System.Web;
using DevExpress.Spreadsheet.Functions;

namespace DevExpress.Web.Demos {
    public class DiscountFunction : ICustomFunction {
        const string functionName = "Discount";
        readonly ParameterInfo[] functionParameters;

        public DiscountFunction() {
            this.functionParameters = new ParameterInfo[] { new ParameterInfo(ParameterType.Value, ParameterAttributes.Required), 
            new ParameterInfo(ParameterType.Value, ParameterAttributes.Required)};
        }

        string IFunction.Name { get { return functionName; } }
        ParameterInfo[] IFunction.Parameters { get { return functionParameters; } }
        ParameterType IFunction.ReturnType { get { return ParameterType.Value; } }
        // Reevaluate cells on every recalculation. 
        // The SpreadsheetControl ignores this value and assumes that the Volatile is always true.
        bool IFunction.Volatile { get { return true; } }

        ParameterValue IFunction.Evaluate(IList<ParameterValue> parameters, EvaluationContext context) {
            double quantity;
            double price;
            ParameterValue quantityParameter;
            ParameterValue priceParameter;

            quantityParameter = parameters[0];
            if(quantityParameter.IsError)
                return quantityParameter;
            else
                quantity = quantityParameter.NumericValue;

            priceParameter = parameters[1];
            if(priceParameter.IsError)
                return priceParameter;
            else
                price = priceParameter.NumericValue;

            if(quantity >= 50)
                return Math.Round(quantity * price * 0.25, 2);
            if(quantity >= 20)
                return Math.Round(quantity * price * 0.1, 2);
            return 0;

        }
        string IFunction.GetName(System.Globalization.CultureInfo culture) {
            return functionName;
        }
    }
}
